go/internal/services/ratelimit/sliding_window.go

```diff
@@ -28,34 +29,91 @@ type service struct {
 	// identifier+sequence -> bucket
 	buckets map[string]*bucket
 
-	peerMu sync.RWMutex
-	peers  map[string]ratelimitv1connect.RatelimitServiceClient
-
+	peerMu               sync.RWMutex
+	peers                map[string]peer
+	replayBuffer         *buffer.Buffer[*ratelimitv1.ReplayRequest]
 	replayCircuitBreaker circuitbreaker.CircuitBreaker[*connect.Response[ratelimitv1.ReplayResponse]]
 }
 
-func (r *service) Ratelimit(ctx context.Context, req RatelimitRequest) RatelimitResponse {
-	ctx, span := tracing.Start(ctx, "slidingWindow.Ratelimit")
+type Config struct {
+	Logger  logging.Logger
+	Cluster cluster.Cluster
+	Clock   clock.Clock
+}
+
+func New(config Config) (*service, error) {
+
+	if config.Clock == nil {
+		config.Clock = clock.New()
+	}
+
+	s := &service{
+		clock:                config.Clock,
+		logger:               config.Logger,
+		cluster:              config.Cluster,
+		shutdownCh:           make(chan struct{}),
+		bucketsLock:          sync.RWMutex{},
+		buckets:              make(map[string]*bucket),
+		peerMu:               sync.RWMutex{},
+		peers:                make(map[string]peer),
+		replayBuffer:         buffer.New[*ratelimitv1.ReplayRequest](10_000, true),
+		replayCircuitBreaker: circuitbreaker.New[*connect.Response[ratelimitv1.ReplayResponse]]("replayRatelimitRequest"),
+	}
```

Comment:
Nit: Why one of the locks called "xxxxLock" and "xxMu". I know `mu` is the convention anyways just a nit. Also why don't we have a lock per bucket probably would a bit faster or lock per peer? We could just have another map of peers and locks or buckets and locks.

Reply:
+1 on naming

we have a lock per bucket
why would we need a lock per peer?